ext/zpages/test/tracez_data_aggregator_benchmark.cc

```diff
@@ -0,0 +1,380 @@
+void StartEndSpans(
+    std::shared_ptr<opentelemetry::trace::Tracer> &tracer,
+    int num_spans,
+    bool is_unique = false)
+{
+  opentelemetry::trace::StartSpanOptions start;
+  opentelemetry::trace::EndSpanOptions end;
+  start.start_steady_time = SteadyTimestamp(nanoseconds(0));
+  end.end_steady_time = SteadyTimestamp(nanoseconds(1));
+  for (; num_spans > 0; num_spans--)
+    tracer->StartSpan(is_unique ? std::to_string(num_spans) : "", start)->End(end);
+}
+
+/*
+ * Helper function that creates and ends num_spans spans instantly, while
+ * simulating different latencies. If is_unique is true, then all spans will
+ * have different names.
+ */
+void StartEndSpansLatency(
+    std::shared_ptr<opentelemetry::trace::Tracer> &tracer,
+    int num_spans,
+    bool is_unique = false)
+{
+  opentelemetry::trace::StartSpanOptions start;
+  start.start_steady_time = SteadyTimestamp(nanoseconds(0));
+  for (; num_spans > 0; num_spans--)
+  {
+    // Latency bucket depends on the index
+    auto latency_band = kLatencyBoundaries[num_spans % kLatencyBoundaries.size()];
+    opentelemetry::trace::EndSpanOptions end;
+    end.end_steady_time = SteadyTimestamp(latency_band);
+
+    tracer->StartSpan(is_unique ? std::to_string(num_spans) : "", start)->End(end);
+  }
+}
+
+/*
+ * Helper function that creates and ends num_spans spans instantly, while
+ * simulating error codes. If is_unique is true, then all spans will have
+ * different names.
+ */
+void StartEndSpansError(
+    std::shared_ptr<opentelemetry::trace::Tracer> &tracer,
+    int num_spans,
+    bool is_unique = false)
+{
+  for (; num_spans > 0; num_spans--)
+    tracer->StartSpan(is_unique ? std::to_string(num_spans) : "")
+        ->SetStatus(opentelemetry::trace::CanonicalCode::CANCELLED, "");
+}
+
+/*
+ * Helper function that creates i spans that always run. To remain in scope and running,
+ * spans are referenced in the passed in vector. If is_unique is true, then all spans
+ * will have different names.
+ */
+void StartSpans(
+    std::vector<opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>> &spans,
+    std::shared_ptr<opentelemetry::trace::Tracer> &tracer,
+    int num_spans,
+    bool is_unique = false)
+{
+  for (; num_spans > 0; num_spans--)
+    spans.push_back(tracer->StartSpan(is_unique ? std::to_string(num_spans) : ""));
+}
+
+/*
+ * Helper function that creates about i spans, evenly split between running, latency (which
+ * is further split by bucket), and error. If is_unique is true, then all spans will have
+ * different names. The running spans vector is returned so that they remain running.
+ */
+std::vector<opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>> MakeManySpans(
+    std::shared_ptr<opentelemetry::trace::Tracer> &tracer,
+    int num_spans,
+    bool is_unique = false)
+{
+  // Running spans must be stored in a vector in order to stay running
+  std::vector<opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>> running_spans;
```

Comment:
i think this might cause a data race in some specific cases since three threads are modifying the vector. might be better to just do it all in one thread since it's not going to be timed by the benchmark anymore with Pause/Resume.

Reply:
the error and latency functions currently make and end spans within their respective functions without passing in the vector, since spans are exported to the processor onend. do you think it should be thread safe in that case?

Reply:
whoops -- you're right! for some reason I thought it was being passed into multiple threads. this is definitely thread-safe then.

Reply:
the lines are long, it's hard to see what's going on at first glance, and my comment wasn't very clear, so thanks for pointing it out and i'll add more context!!